developing/vcs/README.md

```diff
@@ -2,7 +2,321 @@
+by turning on the following settings at a minimum:
+
+  * Require pull request reviews before merging
+  * Require status checks to pass before merging
+  * Include administrators
+
+* Under `Settings->Merge button`:
+  * Uncheck all options except for `Allow merge commits` because it's the only
+  one that allows deleting local branches with `-d`. We should never have to use
+  `-D` to delete a branch.
+
+## PR size
+
+Pull requests should be small and focused enough to be reviewable in under 15,
+or ideally under 5, minutes.
+
+## Git workflow
+
+### Fork vs clone
+
+If it's a repo you have write and push access to, clone it. Otherwise, fork it.
+
+### Cloning with SSH vs HTTPS
+
+Do we have strong opinions on this?
+
+### Merge vs rebase
+
+#### TL;DR
+
+If you care about keeping a clean commit history, where each commit has a great
+message, represents a meaningful change, and passes tests and other checks, you
+should use `rebase`. A history that is easy to read and follow is more valuable
+than one that is hard to understand. The full history remains available via
+`git reflog`.
+
+The closest we can come to the best of both worlds is to use `merge` locally,
+and then squash commits via the GitHub website. This comes with two downsides
+though: we are forced to use `-D` to delete local branches, and we become
+dependent on GitHub.
+
+#### Merge
+
+Pros:
+
+* Safer since you don't have to force push
+* History does not change, and includes the good, bad, and ugly
+
+Cons:
+
+* Noisier history with commits that don't add any value and that
+can cause confusion when reviewing the log
+* Not all commits might pass tests, making debugging hard
+* Cannot use interactive rebase to squash and/or reorder commits
+* Increases risk of losing code when handling conflicts due to working with
+multiple commits, some of which might be broken.
+
+Ways to mitigate cons:
+
+* There is a way to squash commits locally without `rebase`, but it involves
+[several steps](https://gist.github.com/n00neimp0rtant/9515611), the last of
+which requires a force push, which defeats the purpose of using `merge`, and
+does not allow reordering or choosing more than one commit.
+
+* You can use `merge` locally, and then squash and reorder commits on GitHub
+right before merging, but that forces you to delete local branches with `-D`,
+which we don't think anyone should have to do.
+
+#### Rebase
+
+Pros:
+
+* Cleaner, more meaningful, easier to understand commit history
+* Easier to keep only commits that pass tests, which makes debugging
+easier with tools like `git bisect`.
+* Safer and easier to handle conflicts when commits are squashed
+* Easier to revert bad commits
+* Easier to change commit messages in a series of commits. We value great commit
+messages, so it should be easy to improve any message in a series of commits.
+* Allows reordering commits for more meaningful grouping. For example, you might
```

Comment:
I think the most idiomatic way to achieve this is aggressive local branching. But as discussed offline, that's not a universal practice.